# This is the Coursera Capstone for IBM Professional Certification 

# 1.  Web Scrapping and Data Preparation for a Dataset of Top Secondary Schools in Melbourne. 
### 2nd Week Final Code.
    Import required libraries

In [1]:
#import the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# we import the libraries for request and BeautifulSoup
from urllib.request import urlopen
from bs4 import BeautifulSoup


### Define the url variable to access to the Top Secondary Schools in Melbourne Australia.

In [2]:
url = 'https://bettereducation.com.au/school/secondary/vic/melbourne_top_secondary_schools.aspx'
#html = urlopen(url)

### Let's use the pandas' read_html method to get the table from the url given. 

In [3]:
dflst = pd.read_html(url,header =0)
df = dflst[0]
dflst[0].head()

School  Postcode  \
0               Mac.Robertson Girls' High School      3004   
1                          Melbourne High School      3141   
2            Nossal High School,Berwick,VIC,3806      3806   
3     Suzanne Cory High School,Werribee,VIC,3030      3030   
4  Presbyterian Ladies' College,Burwood,VIC,3125      3125   

   State Overall Score  English  Maths  Total Enrolments  Trend / Compare  \
0                  100      NaN    NaN               954  Trend / Compare   
1                  100      NaN    NaN              1357  Trend / Compare   
2                  100      NaN    NaN               832  Trend / Compare   
3                  100      NaN    NaN               871  Trend / Compare   
4                  100      NaN    NaN              1407  Trend / Compare   

           Sector  
0      Government  
1      Government  
2      Government  
3      Government  
4  Non-government

### Let's start to clean the dataframe
Drop the followings columns: English, Maths, Total Enrolments and Trend/Compare

In [4]:
#df = 
melb_schl = df[['School', 'Postcode', 'State Overall Score', 'Total Enrolments','Sector']]
gov_schools = melb_schl[melb_schl['Sector'] == 'Government']
nongov_schools = melb_schl[melb_schl['Sector'] == 'Non-government']
gov_schools.head()
nongov_schools.head()
print(nongov_schools.shape)
print(gov_schools.shape)
print(melb_schl.shape)

(93, 5)
(37, 5)
(130, 5)


In [5]:
melb_schl.head()

School  Postcode  \
0               Mac.Robertson Girls' High School      3004   
1                          Melbourne High School      3141   
2            Nossal High School,Berwick,VIC,3806      3806   
3     Suzanne Cory High School,Werribee,VIC,3030      3030   
4  Presbyterian Ladies' College,Burwood,VIC,3125      3125   

   State Overall Score  Total Enrolments          Sector  
0                  100               954      Government  
1                  100              1357      Government  
2                  100               832      Government  
3                  100               871      Government  
4                  100              1407  Non-government

### Next we import the dataset with the Postcodes of the Victoria state in Australia from https://en.wikipedia.org/wiki/List_of_Melbourne_suburbs. Then stored in Google Sheets. This information was geocoded using Google Sheets as provides more stable requests over large amount of address. This Dataset will be available in Github as well. 

In [6]:
vic_geo = pd.read_csv('Geocode VIC - Sheet1.csv')
vic_geo.head()

Suburb  Postcode State  \
0                     Melbourne CBD      3000   VIC   
1                    East Melbourne      3002   VIC   
2                    West Melbourne      3003   VIC   
3  Melbourne CBD(St Kilda Roadarea)      3004   VIC   
4                       South Wharf      3006   VIC   

                                     Address        Lat         Lng  \
0                     Melbourne CBD,3000,VIC -37.815207  144.963937   
1                    East Melbourne,3002,VIC -37.816144  144.980459   
2                    West Melbourne,3003,VIC -37.811450  144.925397   
3  Melbourne CBD(St Kilda Roadarea),3004,VIC -37.838947  144.975959   
4                       South Wharf,3006,VIC -37.825000  144.952000   

                               Local government area  
0                                  City of Melbourne  
1                                  City of Melbourne  
2                                  City of Melbourne  
3  City of Melbourne(east side to High Street, Pr...  
4             City of Melbourne;City of Port Phillip

In [7]:
vic_geo.shape

(549, 7)

In [8]:
melb_schl[melb_schl['Sector']=="Government"].count()

School                 37
Postcode               37
State Overall Score    37
Total Enrolments       37
Sector                 37
dtype: int64

### Adding Suburb + VIC(state)+postcode the field School

In [9]:
#melb_schl[melb_schl['Sector']=="Government"]

for idx, school, postcode in zip(melb_schl[melb_schl['Sector']=="Government"].index,melb_schl[melb_schl['Sector']=="Government"].School, melb_schl[melb_schl['Sector']=="Government"].Postcode):
    if school == ("Nossal High School,Berwick,VIC,3806"):
        print("School already with Suburb and postcode")
    elif (school == "Suzanne Cory High School,Werribee,VIC,3030"):
        print("School already with Suburb and postcode")
    elif (school == "Auburn High School,Hawthorn East,VIC,3123"):
        print("School already with Suburb and postcode")
    elif school == "Albert Park College,Albert Park,VIC,3206":
        print("School already with Suburb and postcode")
    elif school == "Alamanda K-9 College,Point Cook,VIC,3030":
        print("School already with Suburb and postcode")
    else:
        i = vic_geo[vic_geo['Postcode']== postcode].index.values[0]
        temp_sub = vic_geo.loc[i,'Suburb']
        melb_schl.at[idx,'School'] = school+','+temp_sub+','+'VIC'+','+str(postcode)
        print("New School string is: {}".format(melb_schl.loc[idx,'School']))

New School string is: Mac.Robertson Girls' High School,Melbourne CBD(St Kilda Roadarea),VIC,3004
New School string is: Melbourne High School,South Yarra,VIC,3141
School already with Suburb and postcode
School already with Suburb and postcode
New School string is: Box Hill High School,Box Hill,VIC,3128
New School string is: University High School,Parkville,VIC,3052
New School string is: Glen Waverley Secondary College,Glen Waverley,VIC,3150
New School string is: Balwyn High School,Balwyn North,VIC,3104
New School string is: Victorian College Of The Arts Secondary School,South Wharf,VIC,3006
New School string is: East Doncaster Secondary College,Doncaster East,VIC,3109
New School string is: Melbourne Girls' College,Burnley,VIC,3121
New School string is: Mckinnon Secondary College,Bentleigh,VIC,3204
New School string is: Vermont Secondary College,Vermont,VIC,3133
New School string is: Mount Waverley Secondary College,Mount Waverley,VIC,3149
New School string is: Canterbury Girls Secondary

In [10]:
# Validate the results
print(melb_schl['School'])

0      Mac.Robertson Girls' High School,Melbourne CBD...
1             Melbourne High School,South Yarra,VIC,3141
2                    Nossal High School,Berwick,VIC,3806
3             Suzanne Cory High School,Werribee,VIC,3030
4          Presbyterian Ladies' College,Burwood,VIC,3125
5                  Fintona Girls' School,Balwyn,VIC,3103
6                                     Haileybury College
7            Huntingtower School,Mount Waverley,VIC,3149
8       Korowa Anglican Girls' School,Glen Iris,VIC,3146
9          Camberwell Grammar School,Canterbury,VIC,3126
10           St Leonard's College,Brighton East,VIC,3187
11                              Melbourne Grammar School
12     Camberwell Girls Grammar School,Canterbury,VIC...
13                   Trinity Grammar School,Kew,VIC,3101
14                    St Kevin's College,Toorak,VIC,3142
15                      Scotch College,Hawthorn,VIC,3122
16             Lauriston Girls' School,Armadale,VIC,3143
17     Penleigh & Essendon Gram

## There is some missing data in school columns, that is related to the Suburbs of the dataframe.

### This will be updated manually as there are suburbs in the School column that share same postcode or wrong postcode.

In [11]:
# There is some missing data in school column, that is related to the Suburb in 7 rows.
# Haileybury College, VIC 3173
i = melb_schl[melb_schl['School']== 'Haileybury College'].index.values[0]
temp = melb_schl.loc[i,'School']
melb_schl.at[i,'School'] = temp+','+'Keysborough'

#Suzanne Cory High School,Werribee,VIC,3030
#i = melb_schl[melb_schl['School']== "Suzanne Cory High School,Werribee,VIC,3030,Cocoroc,VIC,3030"].index.values[0]
#melb_schl.at[i,'School'] = "Suzanne Cory High School,Werribee,VIC,3030"

#Nossal High School,Berwick,VIC,3806,Berwick,VIC,3806
#i = melb_schl[melb_schl['School']== "Nossal High School,Berwick,VIC,3806,Berwick,VIC,3806"].index.values[0]
#melb_schl.at[i,'School'] = "Nossal High School,Berwick,VIC,3806"

#Auburn High School,Hawthorn East,VIC,3123,Hawthorn East,VIC,3123
#i = melb_schl[melb_schl['School']== "Auburn High School,Hawthorn East,VIC,3123,Hawthorn East,VIC,3123"].index.values[0]
#melb_schl.at[i,'School'] = "Auburn High School,Hawthorn East,VIC,3123"

#Albert Park College,Albert Park,VIC,3206,Albert Park,VIC,3206
#i = melb_schl[melb_schl['School']== "Albert Park College,Albert Park,VIC,3206,Albert Park,VIC,3206"].index.values[0]
#melb_schl.at[i,'School'] = "Albert Park College,Albert Park,VIC,3206"

#Melbourne Girls' College Richmond 3121
i = melb_schl[melb_schl['School']== "Melbourne Girls' College,Burnley,VIC,3121"].index.values[0]
melb_schl.at[55,'School']=  "Melbourne Girls' College,Richmond,VIC,3121"

#Victorian College Of The Arts Secondary School, Southbank, 3006
i = melb_schl[melb_schl['School']== "Victorian College Of The Arts Secondary School,South Wharf,VIC,3006"].index.values[0]
melb_schl.at[i,'School']= "Victorian College Of The Arts Secondary School,Southbank,VIC,3006"


# Melbourne Grammar School 3004

i = melb_schl[melb_schl['School']=='Melbourne Grammar School'].index.values[0]
temp = melb_schl.loc[i,'School']
melb_schl.at[i,'School']= temp+','+"Melbourne CBD(St Kilda Roadarea)"

#School: Wesley College,Melbourne,VIC,3004
i = melb_schl[melb_schl['School']=='Wesley College,Melbourne,VIC,3004'].index.values[0]
temp = melb_schl.loc[i,'School']
melb_schl.at[i,'School']= temp+','+"Melbourne CBD(St Kilda Roadarea)"



#School: Mckinnon Secondary College at Suburb: McKinnon, postcode 3204
i = melb_schl[melb_schl['School']=='Mckinnon Secondary College,Bentleigh,VIC,3204'].index.values[0]
melb_schl.at[i,'School']= "McKinnon Secondary College,McKinnon,VIC,3204"

#School: Viewbank College,Viewbank,VIC,3084
i = melb_schl[melb_schl['School']=='Viewbank College,Eaglemont,VIC,3084'].index.values[0]
melb_schl.at[i,'School']= "Viewbank College,Viewbank,VIC,3084"


#New School string is: Princes Hill Secondary College,Carlton North,VIC,3054
i = melb_schl[melb_schl['School']=='Princes Hill Secondary College,Carlton North,VIC,3054'].index.values[0]
melb_schl.at[i,'School']= "Princes Hill Secondary College,Princes Hill,VIC,3054" 
    



#117 Virtual School Victoria. As it is distance education has no physical location
#i = melb_schl[melb_schl['School']== "Virtual School Victoria"].index.values[0]
#melb_schl.drop(i, inplace = True)

melb_schl.reset_index(inplace = True,drop =True)
melb_schl

School  Postcode  \
0    Mac.Robertson Girls' High School,Melbourne CBD...      3004   
1           Melbourne High School,South Yarra,VIC,3141      3141   
2                  Nossal High School,Berwick,VIC,3806      3806   
3           Suzanne Cory High School,Werribee,VIC,3030      3030   
4        Presbyterian Ladies' College,Burwood,VIC,3125      3125   
5                Fintona Girls' School,Balwyn,VIC,3103      3103   
6                       Haileybury College,Keysborough      3173   
7          Huntingtower School,Mount Waverley,VIC,3149      3149   
8     Korowa Anglican Girls' School,Glen Iris,VIC,3146      3146   
9        Camberwell Grammar School,Canterbury,VIC,3126      3126   
10         St Leonard's College,Brighton East,VIC,3187      3187   
11   Melbourne Grammar School,Melbourne CBD(St Kild...      3004   
12   Camberwell Girls Grammar School,Canterbury,VIC...      3126   
13                 Trinity Grammar School,Kew,VIC,3101      3101   
14                  St Kevin's College,Toorak,VIC,3142      3142   
15                    Scotch College,Hawthorn,VIC,3122      3122   
16           Lauriston Girls' School,Armadale,VIC,3143      3143   
17   Penleigh & Essendon Grammar School,Keilor East...      3033   
18           Kilvington Grammar School,Ormond,VIC,3204      3204   
19                   Ruyton Girls' School,Kew,VIC,3101      3101   
20               St Margaret's School,Berwick,VIC,3806      3806   
21              Box Hill High School,Box Hill,VIC,3128      3128   
22    Lighthouse Christian College,Cranbourne,VIC,3977      3977   
23           University High School,Parkville,VIC,3052      3052   
24   Strathcona Baptist Girls' Grammar School,Cante...      3126   
25   Glen Waverley Secondary College,Glen Waverley,...      3150   
26   Waverley Christian College,Wantirna South,VIC,...      3152   
27              Tintern Grammar,Ringwood East,VIC,3135      3135   
28          Shelford Girls' Grammar,Caulfield,VIC,3161      3161   
29      Mentone Girls' Grammar School,Mentone,VIC,3194      3194   
..                                                 ...       ...   
100      Alphington Grammar School,Alphington,VIC,3078      3078   
101                  Kew High School,Kew East,VIC,3102      3102   
102                Salesian College,Chadstone,VIC,3148      3148   
103              Avila College,Mount Waverley,VIC,3149      3149   
104            Catholic Ladies College,Eltham,VIC,3095      3095   
105           Buckley Park College,Aberfeldie,VIC,3040      3040   
106           Frankston High School,Frankston,VIC,3199      3199   
107   Mentone Girls Secondary College,Mentone,VIC,3194      3194   
108         Our Lady of Sion College,Box Hill,VIC,3128      3128   
109       Ringwood Secondary College,Ringwood,VIC,3134      3134   
110          Star of the Sea College,Brighton,VIC,3186      3186   
111           Blackburn High School,Blackburn,VIC,3130      3130   
112      Casey Grammar School,Cranbourne East,VIC,3977      3977   
113             Santa Maria College,Northcote,VIC,3070      3070   
114    Plenty Valley Christian College,Doreen,VIC,3754      3754   
115           Alamanda K-9 College,Point Cook,VIC,3030      3030   
116      St Michael's Grammar School,St Kilda,VIC,3182      3182   
117         Virtual School Victoria,Thornbury,VIC,3071      3071   
118                 Marcellin College,Bulleen,VIC,3105      3105   
119                  Siena College,Camberwell,VIC,3124      3124   
120      Balcombe Grammar School,Mount Martha,VIC,3934      3934   
121  Overnewton Anglican Community College,Keilor,V...      3036   
122  South Oakleigh Secondary College,Oakleigh Sout...      3167   
123         Camberwell High School,Canterbury,VIC,3126      3126   
124  Flinders Christian Community College,Tyabb,VIC...      3913   
125   Cheltenham Secondary College,Cheltenham,VIC,3192      3192   
126                 St Bede's College,Mentone,VIC,3194      3194   
127        Academy of Mary Immaculat

## Save dataframe as csv file to storage and then dowloaded

### In the Top_melb_schools.csv file there is address information (Place/venue, suburb, state code(VIC) and postcode) to geocode and get the Lat/long coordinated per school.
### As Google sheets uses an addons to geocodes cells it is more realiable in terms of large addresses information to use it than geopy package that we have been using previously.  Later on, the file geo_top_melb_schools.csv it is imported in the below cell.

In [14]:

#melb_schl['School'].to_csv("Top_melb_schools.csv", index=False)
# Save dataframe as csv file to storage
project.save_data(data=melb_schl[['School']].to_csv(index=False),file_name='Top_melb_schools.csv',overwrite=True)

{'file_name': 'Top_melb_schools.csv',
 'message': 'File saved to project storage.',
 'bucket_name': 'ibmdatasciencecapstoneprojectbatt-donotdelete-pr-oaltfhfcgpezxe',
 'asset_id': 'bf35b485-ada8-4643-9eae-a7f0ed97e6e4'}

### In the next cell we import the dataset with the School information and the Lat/Lng coordinates that comes from the previous step where it was already geocoded.

In [12]:
geo_schl = pd.read_csv('geo_top_melb_schools.csv')
geo_schl.head()

School        Lat         Lng
0  Mac.Robertson Girls' High School,Melbourne CBD... -37.836729  144.971831
1         Melbourne High School,South Yarra,VIC,3141 -37.835388  144.995677
2                Nossal High School,Berwick,VIC,3806 -38.039219  145.336030
3         Suzanne Cory High School,Werribee,VIC,3030 -37.893567  144.700048
4      Presbyterian Ladies' College,Burwood,VIC,3125 -37.848946  145.107125

## Merge of the Lat/Lng dataframe with the Schools dataframe

In [13]:
#schl_vic 
schl_vic = pd.merge(melb_schl,geo_schl)
schl_vic.head(10)

School  Postcode  \
0  Mac.Robertson Girls' High School,Melbourne CBD...      3004   
1         Melbourne High School,South Yarra,VIC,3141      3141   
2                Nossal High School,Berwick,VIC,3806      3806   
3         Suzanne Cory High School,Werribee,VIC,3030      3030   
4      Presbyterian Ladies' College,Burwood,VIC,3125      3125   
5              Fintona Girls' School,Balwyn,VIC,3103      3103   
6                     Haileybury College,Keysborough      3173   
7        Huntingtower School,Mount Waverley,VIC,3149      3149   
8   Korowa Anglican Girls' School,Glen Iris,VIC,3146      3146   
9      Camberwell Grammar School,Canterbury,VIC,3126      3126   

   State Overall Score  Total Enrolments          Sector        Lat  \
0                  100               954      Government -37.836729   
1                  100              1357      Government -37.835388   
2                  100               832      Government -38.039219   
3                  100               871      Government -37.893567   
4                  100              1407  Non-government -37.848946   
5                  100               450  Non-government -37.814764   
6                   99              3927  Non-government -37.992809   
7                   99               697  Non-government -37.876487   
8                   99               670  Non-government -37.860997   
9                   99              1296  Non-government -37.816624   

          Lng  
0  144.971831  
1  144.995677  
2  145.336030  
3  144.700048  
4  145.107125  
5  145.080433  
6  145.144701  
7  145.136121  
8  145.054004  
9  145.066725

In [14]:
schl_vic.shape

(129, 7)

In [15]:
schl_vic[schl_vic['Sector']=='Government'].School.count()


37

In [16]:
schl_vic[schl_vic['Sector']=='Non-government'].School.count()

92

For loop over the gov_schools Dataframe to select the Suburb, postcode and  Lat/Long coordinates from vic_geo Dataframe.

### For loop over the latlon dataframe
We get the postcode from latlon and search for that postcode in the df dataframe, then the index of df is obtained. 
With the index we can update df to add the latitude and longitud information from the latlon dataframe.

for idx, pcode,lat,lon in zip(latlon.index, latlon['Postcode'], latlon['Latitude'], latlon['Longitude']):
    # Getting the index values of df that match the postcod of latlon dataframe. 
    i = df[df['Postcode'] == pcode].index.values
    #print(i.shape[0])
    if i.shape[0] != 0:
        df.at[i,'Latitude'] = lat
        df.at[i,'Longitude'] =lon
        

df.head()

# 2. Explore and Cluster Secondary Schools in Melbourne

Let's install geopy and folium

In [17]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
#pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
#!pip install folium==0.5.0
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Use geopy library to get the latitude and longitude values of Melbourne City.

In [18]:
address = 'Melbourne, VIC,3000'

geolocator = Nominatim(user_agent="melbourne_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Melbourne City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Melbourne City are -37.8042409, 144.941123149578.


#### Create map of Melbourne using latitude and longitude values

In [19]:
# create map of Melbourne using latitude and longitude values
map_melbourne = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, school, score, enrol, sector in zip(schl_vic['Lat'], schl_vic['Lng'], schl_vic['School'], schl_vic['State Overall Score'], schl_vic['Total Enrolments'], schl_vic['Sector']):
    label = 'School: {}, State Overall Score {}, Total Enrolments {}, Sector: {}'.format(school, score, enrol,sector)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_melbourne)  
    
map_melbourne

In [20]:
CLIENT_ID = 'I1MJ32QYFTODVVPW54MTDO4KK44FYJIAJKV0Z1ULVUVSI1G4' # your Foursquare ID
CLIENT_SECRET = 'EHLA5VNBKTF4AZL4TSI5YU441UCMO1IMN4YHUAPRNQSNMBUY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: I1MJ32QYFTODVVPW54MTDO4KK44FYJIAJKV0Z1ULVUVSI1G4
CLIENT_SECRET:EHLA5VNBKTF4AZL4TSI5YU441UCMO1IMN4YHUAPRNQSNMBUY


In [21]:
schl_vic.loc[schl_vic.index.values[3],'School']

'Suzanne Cory High School,Werribee,VIC,3030'

In [22]:
schl_latitude = schl_vic.loc[schl_vic.index.values[0], 'Lat'] # School latitude value
schl_longitude = schl_vic.loc[schl_vic.index.values[0], 'Lng'] # School longitude value

schl_name = schl_vic.loc[schl_vic.index.values[0], 'School'] # School name

print('Latitude and longitude values of {} are {}, {}.'.format(schl_name, 
                                                               schl_latitude, 
                                                               schl_longitude))

Latitude and longitude values of Mac.Robertson Girls' High School,Melbourne CBD(St Kilda Roadarea),VIC,3004 are -37.8367287, 144.97183090000001.


### Now, let's get the top 50 venues that are in Schools within a radius of 1000 meters.
First, let's create the GET request URL. Name your URL url.

In [29]:
# type your answer here
LIMIT =100 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    schl_latitude, 
    schl_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=I1MJ32QYFTODVVPW54MTDO4KK44FYJIAJKV0Z1ULVUVSI1G4&client_secret=EHLA5VNBKTF4AZL4TSI5YU441UCMO1IMN4YHUAPRNQSNMBUY&v=20180605&ll=-37.8367287,144.97183090000001&radius=1000&limit=100'

In [30]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d76f852ad1789002c9beea3'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Current map view',
  'headerFullLocation': 'Current map view',
  'headerLocationGranularity': 'unknown',
  'totalResults': 104,
  'suggestedBounds': {'ne': {'lat': -37.82772869099999,
    'lng': 144.9832054729414},
   'sw': {'lat': -37.845728709000014, 'lng': 144.96045632705864}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '576ca21c498eecbf1b01fab5',
       'name': 'Crux & Co',
       'location': {'address': 'G01, 35 Albert Road',
        'lat': -37.834504818375116,
        'lng': 144.971561652877,
        'labeledLatLngs': [{'label': 'display',
          'lat': -37.

### Explore Schools in Melbourne
Let's use the function used in the previuos lab to get the near by venues for all the Top Secondary Schools in Melbourne

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['School', 
                  'School Latitude', 
                  'School Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [32]:
## To simplify the code

schl_vic.reset_index(drop = True)
schl_venues = getNearbyVenues(names=schl_vic['School'],
                                   latitudes=schl_vic['Lat'],
                                   longitudes=schl_vic['Lng']
                                  )

Mac.Robertson Girls' High School,Melbourne CBD(St Kilda Roadarea),VIC,3004
Melbourne High School,South Yarra,VIC,3141
Nossal High School,Berwick,VIC,3806
Suzanne Cory High School,Werribee,VIC,3030
Presbyterian Ladies' College,Burwood,VIC,3125
Fintona Girls' School,Balwyn,VIC,3103
Haileybury College,Keysborough
Huntingtower School,Mount Waverley,VIC,3149
Korowa Anglican Girls' School,Glen Iris,VIC,3146
Camberwell Grammar School,Canterbury,VIC,3126
St Leonard's College,Brighton East,VIC,3187
Melbourne Grammar School,Melbourne CBD(St Kilda Roadarea)
Camberwell Girls Grammar School,Canterbury,VIC,3126
Trinity Grammar School,Kew,VIC,3101
St Kevin's College,Toorak,VIC,3142
Scotch College,Hawthorn,VIC,3122
Lauriston Girls' School,Armadale,VIC,3143
Penleigh & Essendon Grammar School,Keilor East,VIC,3033
Kilvington Grammar School,Ormond,VIC,3204
Ruyton Girls' School,Kew,VIC,3101
St Margaret's School,Berwick,VIC,3806
Box Hill High School,Box Hill,VIC,3128
Lighthouse Christian College,Cranbourne,

Let's explore the new Dataframe schl_venues

In [33]:
print(schl_venues.shape)
schl_venues.head()

(3649, 7)


School  School Latitude  \
0  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
1  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
2  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
3  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
4  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   

   School Longitude             Venue  Venue Latitude  Venue Longitude  \
0        144.971831         Crux & Co      -37.834505       144.971562   
1        144.971831  The Kettle Black      -37.833992       144.971451   
2        144.971831  Slater St. Bench      -37.839004       144.976725   
3        144.971831   Kebabs R' Yummy      -37.831966       144.969831   
4        144.971831    Tempura Hajime      -37.832900       144.968930   

              Venue Category  
0                       Café  
1                       Café  
2                       Café  
3  Middle Eastern Restaurant  
4        Japanese Restaurant

In [34]:
# Let's count the venues around the schools.
schl_venues.groupby(['School']).count()

School Latitude  \
School                                                                
Academy of Mary Immaculate,Fitzroy,VIC,3065                     100   
Adass Israel School,Elsternwick,VIC,3185                         52   
Alamanda K-9 College,Point Cook,VIC,3030                          8   
Albert Park College,Albert Park,VIC,3206                         89   
Alphington Grammar School,Alphington,VIC,3078                    11   
Auburn High School,Hawthorn East,VIC,3123                        33   
Avila College,Mount Waverley,VIC,3149                            19   
Bacchus Marsh Grammar,Bacchus Marsh,VIC,3340                      2   
Balcombe Grammar School,Mount Martha,VIC,3934                     5   
Balwyn High School,Balwyn North,VIC,3104                         14   
Beth Rivkah Ladies College,St Kilda East,VIC,3183                58   
Bialik College,Hawthorn East,VIC,3123                            29   
Billanook College,Mooroolbark,VIC,3138                            7   
Blackburn High School,Blackburn,VIC,3130                         20   
Box Hill High School,Box Hill,VIC,3128                            9   
Brentwood Secondary College,Glen Waverley,VIC,3150               17   
Brighton Grammar School,Brighton,VIC,3186                        47   
Brighton Secondary College,Brighton East,VIC,3187                13   
Buckley Park College,Aberfeldie,VIC,3040                         12   
Camberwell Girls Grammar School,Canterbury,VIC,...               19   
Camberwell Grammar School,Canterbury,VIC,3126                    20   
Camberwell High School,Canterbury,VIC,3126                       20   
Canterbury Girls Secondary College,Canterbury,V...               13   
Carey Baptist Grammar School,Kew,VIC,3101                        29   
Casey Grammar School,Cranbourne East,VIC,3977                     6   
Catholic Ladies College,Eltham,VIC,3095                          26   
Caulfield Grammar School,St Kilda East,VIC,3183                  66   
Cheltenham Secondary College,Cheltenham,VIC,3192                 16   
Cornish College,Bangholme,VIC,3175                                4   
Donvale Christian College,Donvale,VIC,3111                        3   
ELTHAM College of Education,Research,VIC,3095                     5   
East Doncaster Secondary College,Doncaster East...               26   
Emmaus College,Vermont South,VIC,3133                            16   
Fintona Girls' School,Balwyn,VIC,3103                            24   
Firbank Grammar School,Brighton,VIC,3186                         35   
Flinders Christian Community College,Tyabb,VIC,...                1   
Frankston High School,Frankston,VIC,3199                          6   
Genazzano FCJ College,Kew,VIC,3101                               13   
Glen Eira College,Caulfield East,VIC,3145                        16   
Glen Waverley Secondary College,Glen Waverley,V...               67   
Glenvale School,Yarrambat,VIC,3091                                2   
Haileybury College,Keysborough                                   11   
Heathdale Christian College,Werribee,VIC,3030                    35   
Heatherton Christian College,Clarinda,VIC,3169                    5   
Highvale Secondary College,Glen Waverley,VIC,3150                10   
Huntingtower School,Mount Waverley,VIC,3149                      20   
Ivanhoe Girls' Grammar School,Ivanhoe,VIC,3079                   21   
Ivanhoe Grammar School,Ivanhoe,VIC,3079                          17   
Kew High School,Kew East,VIC,3102                                22   
Killester College,Springvale,VIC,3171                            24   
Kilvington Grammar School,Ormond,VIC,3204                        22   
Kingswood College,Box Hill,VIC,3128                              15   
Koonung Secondary College,Box Hill North,VIC,3129                 7   
Korowa Anglican Girls' School,Glen Iris,VIC,3146                 21   
Lauriston Girls' School,Armadale,VIC,3143                        56   
Leibl

Let's find out how many unique categories can be curated from all the returned venues

In [35]:
print('There are {} uniques categories.'.format(len(schl_venues['Venue Category'].unique())))
schl_venues.head()

There are 281 uniques categories.


School  School Latitude  \
0  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
1  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
2  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
3  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   
4  Mac.Robertson Girls' High School,Melbourne CBD...       -37.836729   

   School Longitude             Venue  Venue Latitude  Venue Longitude  \
0        144.971831         Crux & Co      -37.834505       144.971562   
1        144.971831  The Kettle Black      -37.833992       144.971451   
2        144.971831  Slater St. Bench      -37.839004       144.976725   
3        144.971831   Kebabs R' Yummy      -37.831966       144.969831   
4        144.971831    Tempura Hajime      -37.832900       144.968930   

              Venue Category  
0                       Café  
1                       Café  
2                       Café  
3  Middle Eastern Restaurant  
4        Japanese Restaurant

## Analyze Each School's Suburb

In [36]:
# one hot encoding
schl_onehot = pd.get_dummies(schl_venues[['Venue Category']], prefix="", prefix_sep="")
schl_onehot.head()
# add neighborhood column back to dataframe
schl_onehot['School'] = schl_venues['School'] 
schl_onehot.head()


Afghan Restaurant  African Restaurant  Antique Shop  Aquarium  Arcade  \
0                  0                   0             0         0       0   
1                  0                   0             0         0       0   
2                  0                   0             0         0       0   
3                  0                   0             0         0       0   
4                  0                   0             0         0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Arts & Crafts Store  \
0                 0                       0            0                    0   
1                 0                       0            0                    0   
2                 0                       0            0                    0   
3                 0                       0            0                    0   
4                 0                       0            0                    0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  \
0                 0                   0                      0   
1                 0                   0                      0   
2                 0                   0                      0   
3                 0                   0                      0   
4                 0                   0                      0   

   Auto Dealership  BBQ Joint  Badminton Court  Bagel Shop  Bakery  Bar  \
0                0          0                0           0       0    0   
1                0          0                0           0       0    0   
2                0          0                0           0       0    0   
3                0          0                0           0       0    0   
4                0          0                0           0       0    0   

   Baseball Field  Basketball Court  Beach  Beach Bar  Beer Bar  Beer Garden  \
0               0                 0      0          0         0            0   
1               0                 0      0          0         0            0   
2               0                 0      0          0         0            0   
3               0                 0      0          0         0            0   
4               0                 0      0          0         0            0   

   Beer Store  Big Box Store  Board Shop  Bookstore  Botanical Garden  \
0           0              0           0          0                 0   
1           0              0           0          0                 0   
2           0              0           0          0                 0   
3           0              0           0          0                 0   
4           0              0           0          0                 0   

   Boutique  Bowling Alley  Breakfast Spot  Brewery  Bridal Shop  \
0         0              0               0        0            0   
1         0              0               0        0            0   
2         0              0               0        0            0   
3         0              0               0        0            0   
4         0              0               0        0            0   

   Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  Bus Station  \
0                0             0              0         0            0   
1                0             0              0         0            0   
2                0             0              0         0            0   
3                0             0              0         0            0   
4                0             0              0         0            0   

   Bus Stop  Business Service  Café  Camera Store  Campground  Candy Store  \
0         0                 0     1             0           0            0   
1         0                 0     1             0           0            0   
2         0                 0     1             0           0            0   
3         0                 0     0             0           0            0   
4         0                 0     0            

In [37]:
# move School column to the first column
fixed_columns = [schl_onehot.columns[-1]] + list(schl_onehot.columns[:-1])
schl_onehot = schl_onehot[fixed_columns]
#toronto_onehot.rename(columns = {'Zeighborhood':'Neighborhood'},inplace = True) 

schl_onehot.head()

Zoo Exhibit  Afghan Restaurant  African Restaurant  Antique Shop  Aquarium  \
0            0                  0                   0             0         0   
1            0                  0                   0             0         0   
2            0                  0                   0             0         0   
3            0                  0                   0             0         0   
4            0                  0                   0             0         0   

   Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0       0                 0                       0            0   
1       0                 0                       0            0   
2       0                 0                       0            0   
3       0                 0                       0            0   
4       0                 0                       0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0                    0                 0                   0   
1                    0                 0                   0   
2                    0                 0                   0   
3                    0                 0                   0   
4                    0                 0                   0   

   Australian Restaurant  Auto Dealership  BBQ Joint  Badminton Court  \
0                      0                0          0                0   
1                      0                0          0                0   
2                      0                0          0                0   
3                      0                0          0                0   
4                      0                0          0                0   

   Bagel Shop  Bakery  Bar  Baseball Field  Basketball Court  Beach  \
0           0       0    0               0                 0      0   
1           0       0    0               0                 0      0   
2           0       0    0               0                 0      0   
3           0       0    0               0                 0      0   
4           0       0    0               0                 0      0   

   Beach Bar  Beer Bar  Beer Garden  Beer Store  Big Box Store  Board Shop  \
0          0         0            0           0              0           0   
1          0         0            0           0              0           0   
2          0         0            0           0              0           0   
3          0         0            0           0              0           0   
4          0         0            0           0              0           0   

   Bookstore  Botanical Garden  Boutique  Bowling Alley  Breakfast Spot  \
0          0                 0         0              0               0   
1          0                 0         0              0               0   
2          0                 0         0              0               0   
3          0                 0         0              0               0   
4          0                 0         0              0               0   

   Brewery  Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  \
0        0            0                0             0              0   
1        0            0                0             0              0   
2        0            0                0             0              0   
3        0            0                0             0              0   
4        0            0                0             0              0   

   Bus Line  Bus Station  Bus Stop  Business Service  Café  Camera Store  \
0         0            0         0                 0     1             0   
1         0            0         0                 0     1             0   
2         0            0         0                 0     1             0   
3         0            0         0                 0     0             0   
4         0            0         0                 0     0             0   

   Campground  Candy Store  Car Wash  Ca

In [38]:
schl_onehot.School

0       Mac.Robertson Girls' High School,Melbourne CBD...
1       Mac.Robertson Girls' High School,Melbourne CBD...
2       Mac.Robertson Girls' High School,Melbourne CBD...
3       Mac.Robertson Girls' High School,Melbourne CBD...
4       Mac.Robertson Girls' High School,Melbourne CBD...
5       Mac.Robertson Girls' High School,Melbourne CBD...
6       Mac.Robertson Girls' High School,Melbourne CBD...
7       Mac.Robertson Girls' High School,Melbourne CBD...
8       Mac.Robertson Girls' High School,Melbourne CBD...
9       Mac.Robertson Girls' High School,Melbourne CBD...
10      Mac.Robertson Girls' High School,Melbourne CBD...
11      Mac.Robertson Girls' High School,Melbourne CBD...
12      Mac.Robertson Girls' High School,Melbourne CBD...
13      Mac.Robertson Girls' High School,Melbourne CBD...
14      Mac.Robertson Girls' High School,Melbourne CBD...
15      Mac.Robertson Girls' High School,Melbourne CBD...
16      Mac.Robertson Girls' High School,Melbourne CBD...
17      Mac.Ro

Next, let's group rows by School and by taking the mean of the frequency of occurrence of each category

In [39]:
schl_grouped = schl_onehot.groupby('School').mean().reset_index()
schl_grouped.head()

School  Zoo Exhibit  \
0    Academy of Mary Immaculate,Fitzroy,VIC,3065          0.0   
1       Adass Israel School,Elsternwick,VIC,3185          0.0   
2       Alamanda K-9 College,Point Cook,VIC,3030          0.0   
3       Albert Park College,Albert Park,VIC,3206          0.0   
4  Alphington Grammar School,Alphington,VIC,3078          0.0   

   Afghan Restaurant  African Restaurant  Antique Shop  Aquarium  Arcade  \
0                0.0                 0.0           0.0       0.0     0.0   
1                0.0                 0.0           0.0       0.0     0.0   
2                0.0                 0.0           0.0       0.0     0.0   
3                0.0                 0.0           0.0       0.0     0.0   
4                0.0                 0.0           0.0       0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Arts & Crafts Store  \
0              0.01                    0.01          0.0                 0.01   
1              0.00                    0.00          0.0                 0.00   
2              0.00                    0.00          0.0                 0.00   
3              0.00                    0.00          0.0                 0.00   
4              0.00                    0.00          0.0                 0.00   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  \
0          0.010000               0.000               0.030000   
1          0.000000               0.000               0.000000   
2          0.000000               0.125               0.000000   
3          0.033708               0.000               0.011236   
4          0.000000               0.000               0.000000   

   Auto Dealership  BBQ Joint  Badminton Court  Bagel Shop    Bakery  \
0              0.0   0.000000              0.0    0.010000  0.030000   
1              0.0   0.000000              0.0    0.019231  0.019231   
2              0.0   0.000000              0.0    0.000000  0.000000   
3              0.0   0.011236              0.0    0.000000  0.044944   
4              0.0   0.000000              0.0    0.000000  0.000000   

        Bar  Baseball Field  Basketball Court     Beach  Beach Bar  Beer Bar  \
0  0.070000             0.0               0.0  0.000000        0.0       0.0   
1  0.019231             0.0               0.0  0.000000        0.0       0.0   
2  0.000000             0.0               0.0  0.000000        0.0       0.0   
3  0.000000             0.0               0.0  0.033708        0.0       0.0   
4  0.000000             0.0               0.0  0.000000        0.0       0.0   

   Beer Garden  Beer Store  Big Box Store  Board Shop  Bookstore  \
0         0.01         0.0            0.0         0.0   0.020000   
1         0.00         0.0            0.0         0.0   0.000000   
2         0.00         0.0            0.0         0.0   0.000000   
3         0.00         0.0            0.0         0.0   0.011236   
4         0.00         0.0            0.0         0.0   0.000000   

   Botanical Garden  Boutique  Bowling Alley  Breakfast Spot  Brewery  \
0               0.0       0.0            0.0        0.000000      0.0   
1               0.0       0.0            0.0        0.000000      0.0   
2               0.0       0.0            0.0        0.000000      0.0   
3               0.0       0.0            0.0        0.044944      0.0   
4               0.0       0.0            0.0        0.000000      0.0   

   Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0          0.0              0.0      0.000000            0.0       0.0   
1          0.0              0.0      0.038462            0.0       0.0   
2          0.0              0.0      0.000000            0.0       0.0   
3          0.0              0.0      0.011236            0.0       0.0   
4          0.0              0.0      0.000000            0.0       0.0   

   Bus Station  Bus Stop  Business Service      Café  Camera Store  \
0          0.0       0.0         

In [40]:
schl_grouped.head()

School  Zoo Exhibit  \
0    Academy of Mary Immaculate,Fitzroy,VIC,3065          0.0   
1       Adass Israel School,Elsternwick,VIC,3185          0.0   
2       Alamanda K-9 College,Point Cook,VIC,3030          0.0   
3       Albert Park College,Albert Park,VIC,3206          0.0   
4  Alphington Grammar School,Alphington,VIC,3078          0.0   

   Afghan Restaurant  African Restaurant  Antique Shop  Aquarium  Arcade  \
0                0.0                 0.0           0.0       0.0     0.0   
1                0.0                 0.0           0.0       0.0     0.0   
2                0.0                 0.0           0.0       0.0     0.0   
3                0.0                 0.0           0.0       0.0     0.0   
4                0.0                 0.0           0.0       0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Arts & Crafts Store  \
0              0.01                    0.01          0.0                 0.01   
1              0.00                    0.00          0.0                 0.00   
2              0.00                    0.00          0.0                 0.00   
3              0.00                    0.00          0.0                 0.00   
4              0.00                    0.00          0.0                 0.00   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  \
0          0.010000               0.000               0.030000   
1          0.000000               0.000               0.000000   
2          0.000000               0.125               0.000000   
3          0.033708               0.000               0.011236   
4          0.000000               0.000               0.000000   

   Auto Dealership  BBQ Joint  Badminton Court  Bagel Shop    Bakery  \
0              0.0   0.000000              0.0    0.010000  0.030000   
1              0.0   0.000000              0.0    0.019231  0.019231   
2              0.0   0.000000              0.0    0.000000  0.000000   
3              0.0   0.011236              0.0    0.000000  0.044944   
4              0.0   0.000000              0.0    0.000000  0.000000   

        Bar  Baseball Field  Basketball Court     Beach  Beach Bar  Beer Bar  \
0  0.070000             0.0               0.0  0.000000        0.0       0.0   
1  0.019231             0.0               0.0  0.000000        0.0       0.0   
2  0.000000             0.0               0.0  0.000000        0.0       0.0   
3  0.000000             0.0               0.0  0.033708        0.0       0.0   
4  0.000000             0.0               0.0  0.000000        0.0       0.0   

   Beer Garden  Beer Store  Big Box Store  Board Shop  Bookstore  \
0         0.01         0.0            0.0         0.0   0.020000   
1         0.00         0.0            0.0         0.0   0.000000   
2         0.00         0.0            0.0         0.0   0.000000   
3         0.00         0.0            0.0         0.0   0.011236   
4         0.00         0.0            0.0         0.0   0.000000   

   Botanical Garden  Boutique  Bowling Alley  Breakfast Spot  Brewery  \
0               0.0       0.0            0.0        0.000000      0.0   
1               0.0       0.0            0.0        0.000000      0.0   
2               0.0       0.0            0.0        0.000000      0.0   
3               0.0       0.0            0.0        0.044944      0.0   
4               0.0       0.0            0.0        0.000000      0.0   

   Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0          0.0              0.0      0.000000            0.0       0.0   
1          0.0              0.0      0.038462            0.0       0.0   
2          0.0              0.0      0.000000            0.0       0.0   
3          0.0              0.0      0.011236            0.0       0.0   
4          0.0              0.0      0.000000            0.0       0.0   

   Bus Station  Bus Stop  Business Service      Café  Camera Store  \
0          0.0       0.0         

In [41]:
schl_grouped.shape

(129, 281)

New function that sort the venues in descending order.

In [42]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each School/Suburb using schl_grouped dataframe.

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['School']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
schl_venues_sorted = pd.DataFrame(columns=columns)
schl_venues_sorted['School'] = schl_grouped['School']

for ind in np.arange(schl_grouped.shape[0]):
    schl_venues_sorted.iloc[ind, 1:] = return_most_common_venues(schl_grouped.iloc[ind, :], num_top_venues)

schl_venues_sorted.head()

School 1st Most Common Venue  \
0    Academy of Mary Immaculate,Fitzroy,VIC,3065                  Café   
1       Adass Israel School,Elsternwick,VIC,3185                  Café   
2       Alamanda K-9 College,Point Cook,VIC,3030            Playground   
3       Albert Park College,Albert Park,VIC,3206                  Café   
4  Alphington Grammar School,Alphington,VIC,3078           Golf Course   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0              Wine Bar                        Bar        Ice Cream Shop   
1      Greek Restaurant  Middle Eastern Restaurant  Fast Food Restaurant   
2           Snack Place                Coffee Shop           Pizza Place   
3    Italian Restaurant                     Bakery        Breakfast Spot   
4         Train Station       Gym / Fitness Center  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Deli / Bodega          Cocktail Bar                   Pub   
1      Sushi Restaurant          Burger Joint   Japanese Restaurant   
2                  Park    Athletics & Sports        Farmers Market   
3         Deli / Bodega                 Beach      Asian Restaurant   
4     Convenience Store          Liquor Store           Pizza Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Australian Restaurant           Pizza Place            Coffee Shop  
1     Light Rail Station                  Park  Portuguese Restaurant  
2                  Field     Fish & Chips Shop            Fish Market  
3    Japanese Restaurant                   Gym          Grocery Store  
4                   Park       Thai Restaurant                  Field

## Cluster Neighborhoods
Run k-means to cluster the neighborhood into 4 clusters.

In [44]:
# set number of clusters
kclusters = 4

schl_grouped_clustering = schl_grouped.drop('School', 1)
#Inserting the The Overall State score before clustering, so in the Kmeans, the school score has pondering  values to group the schools
schl_grouped_clustering.insert(1,'Score', schl_vic['State Overall Score']/100)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(schl_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:15]

array([2, 2, 0, 2, 0, 2, 0, 1, 0, 2, 2, 0, 0, 2, 2])

new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [45]:
# add clustering labels
schl_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

schl_venues_sorted.head()

Cluster Labels                                         School  \
0               2    Academy of Mary Immaculate,Fitzroy,VIC,3065   
1               2       Adass Israel School,Elsternwick,VIC,3185   
2               0       Alamanda K-9 College,Point Cook,VIC,3030   
3               2       Albert Park College,Albert Park,VIC,3206   
4               0  Alphington Grammar School,Alphington,VIC,3078   

  1st Most Common Venue 2nd Most Common Venue      3rd Most Common Venue  \
0                  Café              Wine Bar                        Bar   
1                  Café      Greek Restaurant  Middle Eastern Restaurant   
2            Playground           Snack Place                Coffee Shop   
3                  Café    Italian Restaurant                     Bakery   
4           Golf Course         Train Station       Gym / Fitness Center   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Ice Cream Shop         Deli / Bodega          Cocktail Bar   
1  Fast Food Restaurant      Sushi Restaurant          Burger Joint   
2           Pizza Place                  Park    Athletics & Sports   
3        Breakfast Spot         Deli / Bodega                 Beach   
4  Fast Food Restaurant     Convenience Store          Liquor Store   

  7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
0                   Pub  Australian Restaurant           Pizza Place   
1   Japanese Restaurant     Light Rail Station                  Park   
2        Farmers Market                  Field     Fish & Chips Shop   
3      Asian Restaurant    Japanese Restaurant                   Gym   
4           Pizza Place                   Park       Thai Restaurant   

  10th Most Common Venue  
0            Coffee Shop  
1  Portuguese Restaurant  
2            Fish Market  
3          Grocery Store  
4                  Field

### Merging the Suburb Venues with the Schools' data

In [46]:
schl_merged = schl_vic
# merge schl_grouped with schl_vic to add latitude/longitude for each school

schl_merged = schl_merged.join(schl_venues_sorted.set_index('School'), on='School')

schl_merged.head() # check the last columns!

School  Postcode  \
0  Mac.Robertson Girls' High School,Melbourne CBD...      3004   
1         Melbourne High School,South Yarra,VIC,3141      3141   
2                Nossal High School,Berwick,VIC,3806      3806   
3         Suzanne Cory High School,Werribee,VIC,3030      3030   
4      Presbyterian Ladies' College,Burwood,VIC,3125      3125   

   State Overall Score  Total Enrolments          Sector        Lat  \
0                  100               954      Government -37.836729   
1                  100              1357      Government -37.835388   
2                  100               832      Government -38.039219   
3                  100               871      Government -37.893567   
4                  100              1407  Non-government -37.848946   

          Lng  Cluster Labels 1st Most Common Venue    2nd Most Common Venue  \
0  144.971831               2                  Café    Australian Restaurant   
1  144.995677               2                  Café       Italian Restaurant   
2  145.336030               2     Recreation Center             Tennis Court   
3  144.700048               1    Athletics & Sports  Health & Beauty Service   
4  145.107125               2                  Café                      Gym   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0   Japanese Restaurant                 Hotel                   Pub   
1   Japanese Restaurant          Dessert Shop           Coffee Shop   
2               Dog Run                  Café                  Pool   
3  Fast Food Restaurant                 Field     Fish & Chips Shop   
4           Gas Station    Light Rail Station           Coffee Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Coffee Shop    Light Rail Station                   Bar   
1                Bakery                 Hotel         Grocery Store   
2            Food Court     Food & Drink Shop            Food Truck   
3           Fish Market          Fishing Spot           Flea Market   
4    Chinese Restaurant           Pizza Place        Sandwich Place   

  9th Most Common Venue 10th Most Common Venue  
0           Pizza Place  Vietnamese Restaurant  
1             Juice Bar      Korean Restaurant  
2      Football Stadium                   Farm  
3           Flower Shop                    Zoo  
4      Asian Restaurant                    Zoo

In [47]:
#Cleaning some NaN values in the Dataframe.
schl_merged.dropna(inplace = True)
schl_merged.head()

School  Postcode  \
0  Mac.Robertson Girls' High School,Melbourne CBD...      3004   
1         Melbourne High School,South Yarra,VIC,3141      3141   
2                Nossal High School,Berwick,VIC,3806      3806   
3         Suzanne Cory High School,Werribee,VIC,3030      3030   
4      Presbyterian Ladies' College,Burwood,VIC,3125      3125   

   State Overall Score  Total Enrolments          Sector        Lat  \
0                  100               954      Government -37.836729   
1                  100              1357      Government -37.835388   
2                  100               832      Government -38.039219   
3                  100               871      Government -37.893567   
4                  100              1407  Non-government -37.848946   

          Lng  Cluster Labels 1st Most Common Venue    2nd Most Common Venue  \
0  144.971831               2                  Café    Australian Restaurant   
1  144.995677               2                  Café       Italian Restaurant   
2  145.336030               2     Recreation Center             Tennis Court   
3  144.700048               1    Athletics & Sports  Health & Beauty Service   
4  145.107125               2                  Café                      Gym   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0   Japanese Restaurant                 Hotel                   Pub   
1   Japanese Restaurant          Dessert Shop           Coffee Shop   
2               Dog Run                  Café                  Pool   
3  Fast Food Restaurant                 Field     Fish & Chips Shop   
4           Gas Station    Light Rail Station           Coffee Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Coffee Shop    Light Rail Station                   Bar   
1                Bakery                 Hotel         Grocery Store   
2            Food Court     Food & Drink Shop            Food Truck   
3           Fish Market          Fishing Spot           Flea Market   
4    Chinese Restaurant           Pizza Place        Sandwich Place   

  9th Most Common Venue 10th Most Common Venue  
0           Pizza Place  Vietnamese Restaurant  
1             Juice Bar      Korean Restaurant  
2      Football Stadium                   Farm  
3           Flower Shop                    Zoo  
4      Asian Restaurant                    Zoo

 schl_merged = schl_merged.astype({'Cluster Labels':'int32'})

In [48]:
#schl_merged['Cluster Labels'] = toronto_merged.astype({'Cluster Labels':'int32'})
schl_merged['Cluster Labels']

0      2
1      2
2      2
3      1
4      2
5      2
6      0
7      0
8      0
9      2
10     2
11     2
12     2
13     2
14     0
15     0
16     2
17     2
18     0
19     2
20     2
21     2
22     0
23     2
24     2
25     0
26     0
27     0
28     0
29     2
30     2
31     0
32     2
33     2
34     0
35     2
36     2
37     2
38     0
39     2
40     0
41     2
42     2
43     0
44     2
45     2
46     2
47     0
48     0
49     2
50     0
51     2
52     2
53     2
54     0
55     2
56     0
57     2
58     2
59     0
60     0
61     0
62     0
63     0
64     2
65     2
66     0
67     2
68     2
69     2
70     2
71     2
72     2
73     0
74     2
75     2
76     0
77     1
78     2
79     0
80     2
81     2
82     0
83     2
84     2
85     2
86     0
87     2
88     0
89     2
90     0
91     2
92     2
93     0
94     2
95     2
96     0
97     2
98     0
99     3
100    0
101    2
102    0
103    0
104    0
105    2
106    0
107    2
108    0
109    0
110    2
1

### Let's visualize the resulting clusters

In [49]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat,lon,school,score,cluster,sector in zip(schl_merged['Lat'], schl_merged['Lng'], schl_merged['School'],schl_merged['State Overall Score'],schl_merged['Cluster Labels'], schl_merged['Sector']):
    label = folium.Popup(str(school) +', Score '+str(score)+', '+str(sector) +' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine the Cluster given by the Map.

Cluster 1

In [50]:
schl_merged.loc[schl_merged['Cluster Labels'] == 0, schl_merged.columns[[1] + list(range(2, schl_merged.shape[1]))]]

Postcode  State Overall Score  Total Enrolments          Sector  \
6        3173                   99              3927  Non-government   
7        3149                   99               697  Non-government   
8        3146                   99               670  Non-government   
14       3142                   99              2065  Non-government   
15       3122                   99              1866  Non-government   
18       3204                   99               734  Non-government   
22       3977                   98               944  Non-government   
25       3150                   98              1951      Government   
26       3152                   98              1888  Non-government   
27       3135                   98               804  Non-government   
28       3161                   98               526  Non-government   
31       3030                   98              1408  Non-government   
34       3152                   98               620  Non-government   
38       3142                   97               686  Non-government   
40       3111                   97              1314  Non-government   
43       3029                   97              1500  Non-government   
47       3123                   97               867  Non-government   
48       3152                   97               593  Non-government   
50       3109                   96              1618      Government   
54       3930                   96               763  Non-government   
56       3204                   96              2124      Government   
59       3023                   96               634  Non-government   
60       3131                   96               180  Non-government   
61       3133                   96              1489      Government   
62       3149                   95              1887      Government   
63       3079                   95              2089  Non-government   
66       3128                   95               578  Non-government   
73       3101                   95               234  Non-government   
76       3138                   94               680  Non-government   
79       3116                   94               821  Non-government   
82       3169                   94               396  Non-government   
86       3084                   93              1362      Government   
88       3129                   93               947      Government   
90       3171                   93               905  Non-government   
93       3150                   93              1016      Government   
96       3111                   93              1197  Non-government   
98       3175                   92               634  Non-government   
100      3078                   92               515  Non-government   
102      3148                   92              1057  Non-government   
103      3149                   92              1043  Non-government   
104      3095                   92               814  Non-government   
106      3199                   92              1833      Government   
108      3128                   92               940  Non-government   
109      3134                   92              1569      Government   
112      3977                   91               855  Non-government   
115      3030                   91              2408      Government   
117      3071                   91              3445      Government   
118      3105                   91              1432  Non-government   
120      3934                   91               829  Non-government   
121      3036                   91              2098  Non-government   
122      3167                   91               508      Government   
124      3913                   90              1737  Non-government   
125      3192                   90               910      Government   
128      3133                   90              1296  Non-government   

           Lat         Lng  Cluster Labels  

### Let's try to get some insights of each cluster. 

In [51]:
cluster1 = schl_merged.loc[schl_merged['Cluster Labels'] == 0, schl_merged.columns[[1] + list(range(2, schl_merged.shape[1]))]]
len(cluster1['5th Most Common Venue'].unique())

39

In [55]:
cluster1.shape

(54, 17)

In [84]:
cluster1.head()

Postcode  State Overall Score  Total Enrolments          Sector  \
6       3173                   99              3927  Non-government   
7       3149                   99               697  Non-government   
8       3146                   99               670  Non-government   
14      3142                   99              2065  Non-government   
15      3122                   99              1866  Non-government   

          Lat         Lng  Cluster Labels 1st Most Common Venue  \
6  -37.992809  145.144701               0           Golf Course   
7  -37.876487  145.136121               0           Pizza Place   
8  -37.860997  145.054004               0                  Park   
14 -37.836002  145.025030               0                  Park   
15 -37.834231  145.029162               0    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue  4th Most Common Venue  \
6                Stadium         Ice Cream Shop       Greek Restaurant   
7                   Café    Japanese Restaurant  Australian Restaurant   
8                   Café  Vietnamese Restaurant          Train Station   
14         Train Station            Supermarket     Athletics & Sports   
15                  Park          Train Station           Burger Joint   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6      Convenience Store     Electronics Store  Fast Food Restaurant   
7            Supermarket    Chinese Restaurant         Shopping Mall   
8     Light Rail Station                Bakery       Thai Restaurant   
14        Tennis Stadium        Baseball Field              Pharmacy   
15    Athletics & Sports     Fish & Chips Shop        Tennis Stadium   

            8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6   Paper / Office Supplies Store    Athletics & Sports                    Zoo  
7                 Thai Restaurant    Seafood Restaurant                   Park  
8                    Tram Station     Fish & Chips Shop            Pizza Place  
14                   Tram Station        History Museum     Light Rail Station  
15                  Grocery Store        History Museum           Tram Station

In [94]:
cluster1['1st Most Common Venue'].value_counts().head(10)

Park                    8
Golf Course             5
Chinese Restaurant      4
Japanese Restaurant     4
Pizza Place             4
Fast Food Restaurant    4
Café                    3
Electronics Store       2
Stadium                 2
Supermarket             2
Name: 1st Most Common Venue, dtype: int64

In [59]:
cluster1['2nd Most Common Venue'].value_counts().head(10)

Café                     6
Fast Food Restaurant     4
Pizza Place              3
Zoo                      2
Train Station            2
Portuguese Restaurant    2
Athletics & Sports       2
Park                     2
Supermarket              2
Fish & Chips Shop        2
Name: 2nd Most Common Venue, dtype: int64

In [122]:
cluster1['Sector'].describe()

count                 54
unique                 2
top       Non-government
freq                  40
Name: Sector, dtype: object

In [81]:
cluster1[['State Overall Score']].describe()

State Overall Score
count            54.000000
mean             94.648148
std               2.947064
min              90.000000
25%              92.000000
50%              95.000000
75%              97.000000
max              99.000000

In [101]:
# Score of the Government Schools in Cluster 1. 
cluster1.loc[cluster1.Sector == 'Government', cluster1.columns[1]].describe()

count    14.000000
mean     93.357143
std       2.437121
min      90.000000
25%      91.250000
50%      93.000000
75%      95.750000
max      98.000000
Name: State Overall Score, dtype: float64

In [97]:
# Score of the Non-Government Schools in Cluster 1. 
cluster1.loc[cluster1.Sector == 'Non-government', cluster1.columns[1]].describe()

count    40.000000
mean     95.100000
std       3.002563
min      90.000000
25%      92.000000
50%      95.500000
75%      98.000000
max      99.000000
Name: State Overall Score, dtype: float64

It is appreciated a better performance of the Non-government schools in terms of the Overall Score. 
      Gov    Non-gov
Mean  93.36  95.1

For the government sector 75% percent of the schools are above 91.25 of the state overall score, while the non-government ones are above 92 in the overall score.

In [57]:
cluster1['Total Enrolments'].describe()

count      54.000000
mean     1220.425926
std       738.171443
min       180.000000
25%       688.750000
50%       945.500000
75%      1605.750000
max      3927.000000
Name: Total Enrolments, dtype: float64

In [102]:
# Total Enrolments of the Non-Government Schools in Cluster 1. 
cluster1.loc[cluster1.Sector == 'Non-government', cluster1.columns[2]].describe()

count      40.000000
mean     1070.900000
std       686.204716
min       180.000000
25%       661.000000
50%       842.000000
75%      1337.500000
max      3927.000000
Name: Total Enrolments, dtype: float64

In [108]:
# Total Enrolments of the Non-Government Schools in Cluster 1. 
total_nongov = cluster1.loc[cluster1.Sector == 'Non-government', cluster1.columns[2]].sum()
total_nongov

42836

In [103]:
# Total Enrolments of the Government Schools in Cluster 1. 
cluster1.loc[cluster1.Sector == 'Government', cluster1.columns[2]].describe()

count      14.000000
mean     1647.642857
std       737.241491
min       508.000000
25%      1102.500000
50%      1593.500000
75%      1935.000000
max      3445.000000
Name: Total Enrolments, dtype: float64

In [109]:
# Total Enrolments of the Government Schools in Cluster 1. 
total_gov = cluster1.loc[cluster1.Sector == 'Government', cluster1.columns[2]].sum()
total_gov

23067

In [ ]:
#Percentage of the students from the goverment sector

In [127]:
print('Percentage of the Government Schools in cluster 1:%8.2f' % (100*total_gov/(total_gov+total_nongov)))

Percentage of the Government Schools in cluster 1:   35.00


In [128]:
print('Percentage of the Non-government Schools in cluster 1:%8.2f' % (100*total_nongov/(total_gov+total_nongov)))

Percentage of the Non-government Schools in cluster 1:   65.00


In [119]:
print('Total number of students in cluster 1 is {}'.format((total_gov+total_nongov)))

Total number of students in cluster 1 is 65903


In the government schools is observed more density in the enrolments with a mean of 1647.64 compare with 1070 of the non-government ones. The percentile are also showing more crowed schools in the public sector, where at least the half of the schools have around 1593.5 students compare with only 842 from non-government peers. From the total enrolment of the cluster 1, non-goverment institutions have almost 65% of the sample compare with the 35 % of the governement's schools. 


Cluster 2

In [60]:
schl_merged.loc[schl_merged['Cluster Labels'] == 1, schl_merged.columns[[1] + list(range(5, schl_merged.shape[1]))]]

Postcode        Lat         Lng  Cluster Labels 1st Most Common Venue  \
3       3030 -37.893567  144.700048               1    Athletics & Sports   
77      3340 -37.694720  144.430046               1         Event Service   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3   Health & Beauty Service  Fast Food Restaurant                 Field   
77       Athletics & Sports                   Zoo  Fast Food Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3      Fish & Chips Shop           Fish Market          Fishing Spot   
77     Fish & Chips Shop           Fish Market          Fishing Spot   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
3            Flea Market           Flower Shop                    Zoo  
77           Flea Market           Flower Shop      Food & Drink Shop

Cluster 3

In [61]:
schl_merged.loc[schl_merged['Cluster Labels'] == 2, schl_merged.columns[[1] + list(range(2, schl_merged.shape[1]))]]

Postcode  State Overall Score  Total Enrolments          Sector  \
0        3004                  100               954      Government   
1        3141                  100              1357      Government   
2        3806                  100               832      Government   
4        3125                  100              1407  Non-government   
5        3103                  100               450  Non-government   
9        3126                   99              1296  Non-government   
10       3187                   99              1442  Non-government   
11       3004                   99              1799  Non-government   
12       3126                   99               721  Non-government   
13       3101                   99              1443  Non-government   
16       3143                   99               950  Non-government   
17       3033                   99              2587  Non-government   
19       3101                   99               858  Non-government   
20       3806                   99               764  Non-government   
21       3128                   98              1319      Government   
23       3052                   98              1437      Government   
24       3126                   98               736  Non-government   
29       3194                   98               737  Non-government   
30       3134                   98              1249  Non-government   
32       3104                   98              2183      Government   
33       3143                   98               569  Non-government   
35       3101                   97              2014  Non-government   
36       3079                   97               895  Non-government   
37       3101                   97              2344  Non-government   
39       3141                   97               963  Non-government   
41       3184                   97               154  Non-government   
42       3006                   97               406      Government   
44       3185                   97              3114  Non-government   
45       3004                   97              3109  Non-government   
46       3185                   97               478  Non-government   
49       3146                   96               717  Non-government   
51       3101                   96               966  Non-government   
52       3125                   96              1284  Non-government   
53       3101                   96              1943  Non-government   
55       3121                   96              1465      Government   
57       3040                   96               794  Non-government   
58       3142                   96              1224  Non-government   
64       3126                   95              1016      Government   
65       3183                   95               591  Non-government   
67       3170                   95              1395  Non-government   
68       3186                   95              1131  Non-government   
69       3162                   95               572  Non-government   
70       3091                   95               555  Non-government   
71       3194                   95              1522  Non-government   
72       3186                   95              1268  Non-government   
74       3187                   95              1261      Government   
75       3123                   94               413      Government   
78       3070                   94              1811      Government   
80       3206                   94              1265      Government   
81       3166                   94              1005  Non-government   
83       3930                   94              1290  Non-government   
84       3150                   94              1592      Government   
85       3183                   93               427  Non-government   
87       3054                   93               884      Government   
89       3136                   93           

In [63]:
cluster3 = schl_merged.loc[schl_merged['Cluster Labels'] == 2, schl_merged.columns[[1] + list(range(2, schl_merged.shape[1]))]]

In [64]:
cluster3['1st Most Common Venue'].value_counts()

Café                       59
Fast Food Restaurant        2
Park                        2
Light Rail Station          1
Recreation Center           1
Fish & Chips Shop           1
Health & Beauty Service     1
Furniture / Home Store      1
Shopping Mall               1
Grocery Store               1
Gym / Fitness Center        1
Playground                  1
Name: 1st Most Common Venue, dtype: int64

In [66]:
cluster3['2nd Most Common Venue'].value_counts().head(10)

Park                   7
Café                   7
Japanese Restaurant    6
Bar                    5
Coffee Shop            4
Italian Restaurant     4
Pub                    3
Greek Restaurant       3
Pizza Place            3
Light Rail Station     3
Name: 2nd Most Common Venue, dtype: int64

In [70]:
cluster3['Sector'].describe()

count                 72
unique                 2
top       Non-government
freq                  50
Name: Sector, dtype: object

In [67]:
cluster3['State Overall Score'].describe()

count     72.00000
mean      95.50000
std        2.90216
min       90.00000
25%       93.00000
50%       96.00000
75%       98.00000
max      100.00000
Name: State Overall Score, dtype: float64

In [120]:
# Score of the Government Schools in Cluster 3. 
cluster3.loc[cluster3.Sector == 'Government', cluster3.columns[1]].describe()

count     22.000000
mean      94.909091
std        3.053633
min       90.000000
25%       92.250000
50%       94.000000
75%       97.750000
max      100.000000
Name: State Overall Score, dtype: float64

In [121]:
# Score of the Non-Government Schools in Cluster 3. 
cluster3.loc[cluster3.Sector == 'Non-government', cluster3.columns[1]].describe()

count     50.000000
mean      95.760000
std        2.825251
min       90.000000
25%       94.000000
50%       96.000000
75%       98.000000
max      100.000000
Name: State Overall Score, dtype: float64

It is appreciated a better performance of the Non-government schools in terms of the Overall Score. 
      Gov    Non-gov
Mean  94.9  95.76

For the government sector 75% percent of the schools are above 92.25 of the state overall score, while the non-government ones are above 94 in the overall score. 

In [68]:
cluster3['Total Enrolments'].describe()

count      72.000000
mean     1156.944444
std       576.190596
min       154.000000
25%       743.000000
50%      1141.000000
75%      1398.000000
max      3114.000000
Name: Total Enrolments, dtype: float64

In [123]:
# Total Enrolments of the Non-Government Schools in Cluster 3. 
cluster3.loc[cluster3.Sector == 'Non-government', cluster3.columns[2]].describe()

count      50.000000
mean     1137.820000
std       638.977426
min       154.000000
25%       721.250000
50%       964.500000
75%      1370.250000
max      3114.000000
Name: Total Enrolments, dtype: float64

In [124]:
# Total Enrolments of the Government Schools in Cluster 3. 
cluster3.loc[cluster3.Sector == 'Government', cluster3.columns[2]].describe()

count      22.000000
mean     1200.409091
std       408.570290
min       406.000000
25%       963.750000
50%      1242.500000
75%      1417.000000
max      2183.000000
Name: Total Enrolments, dtype: float64

In [136]:
# Total Enrolments of the Government Schools in Cluster 3. 
total_gov3 = cluster3.loc[cluster3.Sector == 'Government', cluster3.columns[2]].sum()
total_gov3

26409

In [137]:
# Total Enrolments of the Non-Government Schools in Cluster 3. 
total_nongov3 = cluster3.loc[cluster3.Sector == 'Non-government', cluster3.columns[2]].sum()
total_nongov3

56891

In [138]:
print('Percentage of the Government Schools in cluster 3:%8.2f' % (100*total_gov3/(total_gov3+total_nongov3)))

Percentage of the Government Schools in cluster 3:   31.70


In [139]:
print('Percentage of the Non-Government Schools in cluster 3:%8.2f' % (100*total_nongov3/(total_gov3+total_nongov3)))

Percentage of the Non-Government Schools in cluster 3:   68.30


In the government schools is observed almost same density in the enrolments with a mean of 1200.409 compare with 1137.82 of the non-government ones. The percentile are also showing more crowed schools in the public sector, where at least the half of the schools have around 1242.5 students compare with only 964.5 from non-government peers. From the total enrolment of the cluster 3, non-goverment institutions have almost 68.3% of the sample compare with the 31.7% of the governement's schools. 

The schools in this cluster are more dense geographically with 68.3% of top schools. 


In [141]:
print('Total number of students in cluster 3 is {}'.format((total_gov3+total_nongov3)))

Total number of students in cluster 3 is 83300


Cluster 4

In [47]:
schl_merged.loc[schl_merged['Cluster Labels'] == 3, schl_merged.columns[[1] + list(range(5, schl_merged.shape[1]))]]

Postcode        Lat         Lng  Cluster Labels 1st Most Common Venue  \
99      3911 -38.192779  145.175103               3          Liquor Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
99             Gift Shop                   Zoo           Flower Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
99                 Field     Fish & Chips Shop           Fish Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
99          Fishing Spot           Flea Market      Food & Drink Shop

Cluster 5

In [57]:
schl_merged.loc[schl_merged['Cluster Labels'] == 4, schl_merged.columns[[1] + list(range(5, schl_merged.shape[1]))]]

Empty DataFrame
Columns: [Postcode, Lat, Lng, Cluster Labels, 1st Most Common Venue, 2nd Most Common Venue, 3rd Most Common Venue, 4th Most Common Venue, 5th Most Common Venue, 6th Most Common Venue, 7th Most Common Venue, 8th Most Common Venue, 9th Most Common Venue, 10th Most Common Venue]
Index: []

### Observation form the Map and the cluster most common Venue

The cluster 2 gather most of the suburbs in Toronto,that represents the ammenities in common for most of the Neighborhoods across the entire Toronto Metro area. 
The Cluster 3 seems to be scatter across the second tier of the city just behind the cbd toward the northeast in between two main roads or train lines from west to east. 
Clusters 1, 4 and 5 are just one Neighborhood per cluster. 

